napari/utils/action_manager.py

```diff
@@ -0,0 +1,328 @@
+
+        button.destroyed.connect(lambda: self._buttons[name].remove(button))
+
+        self._buttons[name].append(button)
+        self._update_gui_elements(name)
+
+    def bind_shortcut(self, name, shortcut):
+        """
+        bind shortcut `shortcut` to trigger action `name`
+        """
+        self._shortcuts[name] = shortcut
+        self._update_shortcut_bindings(name)
+        self._update_gui_elements(name)
+
+    def bind_qaction(self, name, qaction):
+        """
+        Bind the given qaction to an action.
+
+        This will also update the description and shortcut when those changes.
+
+        Same as for shortcut; make ui element flash when action is triggered.
+        """
+        self._qactions[name] = qaction
+        action = self._actions[name]
+
+        def flash(*args):
+            call_with_context(action.command, self.context)
+
+            for b in self._buttons.get(name, []):
+                b._flash_animation.start()
+
+        qaction.destroyed.connect(lambda: self._qactions.pop(name, None))
+        qaction.triggered.connect(flash)
+        qaction.triggered.connect(lambda: self.push(name))
+        self._update_gui_elements(name)
+
+    def unbind_shortcut(self, name):
+        """
+        unbind shortcut for action name
+        """
+        action = self._actions[name]
+        sht = self._shortcuts.get(name)
+        if hasattr(action.keymapprovider, 'bind_key'):
+            action.keymapprovider.bind_key(sht)(None)
+        del self._shortcuts[name]
+        self._update_gui_elements(name)
+
+    def play(self, seq):
+        for i, action_name in enumerate(seq):
+            act = self._actions[action_name]
+
+            for b in self._buttons.get(action_name, []):
+
+                def loc(cmd, b):
+                    def inner():
+                        b._flash_animation.start()
+                        call_with_context(cmd, self.context)
+
+                    return inner
+
+                print('scheduling in', i * 950, act.command, i)
+                from qtpy import QtCore
+
+                QtCore.QTimer.singleShot(i * 950, loc(act.command, b))
+
+
+action_manager = ActionManager()
+
+
+_actions = {
+    'V': 'toggle_selected_visibility',
+    'Control-G': 'toggle_grid',
+    'Shift-Down': 'also_select_layer_below',
+    'Shift-Up': 'also_select_layer_above',
+    'Down': 'select_layer_below',
+    'Up': 'select_layer_above',
+    'Control-A': 'select_all',
+    'Control-T': 'transpose_axes',
+    'Control-R': 'reset_view',
+    'Control-Shift-C': 'toggle_console_visibility',
+    'Control-E': 'roll_axes',
+    'Alt-Down': 'focus_axes_down',
+    'Alt-Up': 'focus_axes_up',
+    'Right': 'increment_dims_right',
+    'Left': 'increment_dims_left',
+    'Control-Y': 'toggle_ndisplay',
+    'Control-Shift-Backspace': 'remove_all_layers',
+    'Control-Shift-Delete': 'remove_all_layers',
+    'Control-Backspace': 'remove_selected',
+    'Control-Delete': 'remove_selected',
+}
```

Comment:
As above... Here I would really like to see the names be uniquified/disambiguated somehow... Fully qualified function/method names seems like a good approach...?

Reply:
This is something that we have elaborated a bit in other meetings, there are two layers here:

1. Who registers the action and
2. under `when` conditions does the action apply?

For 1, I agree with @Carreau on using `plugin_name:action_name`

For 2 it will actually depend on the `when`, since two different (and unique) actions could use the exact same shortcut under a different `when`.

The `when` I am talking about is related more closely to what VSCode displays on the shortcuts
<img width="1311" alt="Screen Shot 2021-04-26 at 3 21 50 PM" src="https://user-images.githubusercontent.com/3627835/116145592-21177e80-a6a3-11eb-8bfc-d8e0e144c56d.png">

We could eventually generalize this as the context, where context are basically callbacks that return either True or False and can be concatenated.